extensions/amp-fit-text/0.1/amp-fit-text.js

```diff
@@ -71,38 +75,17 @@ class AmpFitText extends AMP.BaseElement {
 
   /** @override */
   buildCallback() {
-    this.content_ = this.element.ownerDocument.createElement('div');
-    applyFillContent(this.content_);
-    this.content_.classList.add('i-amphtml-fit-text-content');
-    setStyles(this.content_, {zIndex: 2, visibility: 'hidden'});
-
-    this.contentWrapper_ = this.element.ownerDocument.createElement('div');
-    setStyles(this.contentWrapper_, {lineHeight: `${LINE_HEIGHT_EM_}em`});
-    this.content_.appendChild(this.contentWrapper_);
-
-    this.measurer_ = this.element.ownerDocument.createElement('div');
-    // Note that "measurer" cannot be styled with "bottom:0".
-    setStyles(this.measurer_, {
-      position: 'absolute',
-      top: 0,
-      left: 0,
-      zIndex: 1,
-      visibility: 'hidden',
-      lineHeight: `${LINE_HEIGHT_EM_}em`,
-    });
-
-    realChildNodes(this.element).forEach((node) => {
-      this.contentWrapper_.appendChild(node);
-    });
-    this.updateMeasurerContent_();
-    this.element.appendChild(this.content_);
-    this.element.appendChild(this.measurer_);
+    const {element} = this;
+    buildDom(element.ownerDocument, element);
+    this.content_ = getDescendentByClass(element, CONTENT_CLASS);
+    this.contentWrapper_ = getDescendentByClass(element, CONTENT_WRAPPER_CLASS);
+    this.measurer_ = getDescendentByClass(element, MEASURER_CLASS);
```

Comment:
What if this could look like:

```suggestion
    const {content, contentWrapper, measurer} = buildDom(element.ownerDocument, element);
    this.content_ = content;
    this.contentWrapper_ = contentWrapper;
    this.measurer_ = measurer;
```

avoiding all the DOM reads

Reply:
(Compiler would completely ignore the return value)

Reply:
This idea works really well for classic, but wouldn't work for Bento.
That said, this is a classic component.

I'll test it out!

Reply:
Oh right, just remembered why I didn't pursue this idea.
`buildDom` is only called during client render and skipped for server-render.

So technically the optimal code would get the references from buildDom like you say during csr and then use querySelector for ssr path. I'd personally prefer we keep it simpler for now and opt for the uniform (slow) path in both cases since this is very far away from being the expensive piece of AMP (afaik)

Reply:
Copying from our private thread:
Leaning into the "user over developer" mantra, I think that means we should do something along the lines of
```javascript
/** @type {{foo: !Element, bar: !Element}} */
let AmpFitTextBuiltElements;

/** @returns {!AmpFitTextBuiltElements} */
buildDom(el, doc) {...}
getBuiltElements(el) {...}

buildCallback() {
  /** @type AmpFitTextBuiltElements */
  let builtElements;
  if (!this.element.hasAttribute('i-amphtml-ssr')) {
    builtElements = getBuiltElements(...);
  } else {
    builtElements = buildDom(...);
  }
  
  const {foo, bar} = builtElements;
  this.foo_ = foo;
  this.bar_ = bar;
}
```

It's not a significant amount of extra work but esp for larger elements (I'm thinking carousels that initialize arrays of elements) could avoid a ton of queries on uncompiled pages.

Reply:
I performed some local ad-hoc measurements to get a feel for how expensive hydration and whether it was worth the complexity to avoid it during client-side render (https://github.com/ampproject/amphtml/pull/35101#discussion_r663161558)

The answer was that it is definitely worth it. I'll make this change